File: 03-Maximum-Subarray/maxSubArray.go
package main

import "fmt"

// func maxSubArray(nums []int) int {
// 	max := nums[0]
// 	sum := 0
// 	for _, n := range nums {
// 		if sum < 0 {
// 			sum = n
// 		} else {
// 			sum += n
// 		}
// 		if sum > max {
// 			max = sum
// 		}
// 	}
// 	return max
// }

func maxSubArray(nums []int) int {
	max := nums[0]
	for i := range nums {
		sum := 0
		for _, v := range nums[i:] {
			sum += v
			if sum > max {
				max = sum
			}
		}
	}
	return max
}

func main() {
	// nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	var nums []int = []int{-2, -3 -1}
	res := maxSubArray(nums)
	fmt.Printf("maxSubArray: %d\n", res)
}

File: 03-Maximum-Subarray/maxSubArray.cpp
#include <vector>
#include <iostream>

int maxSubArray(std::vector<int>& nums){
    int max = nums[0];
    for (int i = 0; i < nums.size(); i++)
    {
        int sum = 0;
        for (int j = i; j < nums.size(); j++)
        {
            sum += nums[j];
            if(sum > max)
                max = sum;
        }
    }
    return max;
}

int main(int argc, char const *argv[])
{
    std::vector<int> nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    int res = maxSubArray(nums);
    std::cout << "maxSubArray:" << res << std::endl;
    return 0;
}
